### PR TITLE
python311Packages.jupyterhub: 1.5.0 -> 4.0.0

### PR DESCRIPTION
**Pull request by fabaff**
Changelog: https://github.com/jupyterhub/jupyterhub/blob/4.0.0/docs/source/changelog.md

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 235828` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>28 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.batchspawner</li>
    <li>python310Packages.batchspawner.dist</li>
    <li>python310Packages.dockerspawner</li>
    <li>python310Packages.dockerspawner.dist</li>
    <li>python310Packages.jupyterhub</li>
    <li>python310Packages.jupyterhub-ldapauthenticator</li>
    <li>python310Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python310Packages.jupyterhub-systemdspawner</li>
    <li>python310Packages.jupyterhub-systemdspawner.dist</li>
    <li>python310Packages.jupyterhub-tmpauthenticator</li>
    <li>python310Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python310Packages.jupyterhub.dist</li>
    <li>python310Packages.oauthenticator</li>
    <li>python310Packages.oauthenticator.dist</li>
    <li>python311Packages.batchspawner</li>
    <li>python311Packages.batchspawner.dist</li>
    <li>python311Packages.dockerspawner</li>
    <li>python311Packages.dockerspawner.dist</li>
    <li>python311Packages.jupyterhub</li>
    <li>python311Packages.jupyterhub-ldapauthenticator</li>
    <li>python311Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python311Packages.jupyterhub-systemdspawner</li>
    <li>python311Packages.jupyterhub-systemdspawner.dist</li>
    <li>python311Packages.jupyterhub-tmpauthenticator</li>
    <li>python311Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python311Packages.jupyterhub.dist</li>
    <li>python311Packages.oauthenticator</li>
    <li>python311Packages.oauthenticator.dist</li>
  </ul>
</details>

**Comment by natsukium** [comment]
It seems that sqlalchemy2 is already supported, so it may be possible to build it successfully.
https://github.com/jupyterhub/jupyterhub/pull/4302

**Comment by mweinelt** [comment]
> This pull request first appeared in [4.0.0b1](https://github.com/jupyterhub/jupyterhub/releases/tag/4.0.0b1)

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 235828` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li>python310Packages.batchspawner</li>
    <li>python310Packages.batchspawner.dist</li>
    <li>python310Packages.jupyterhub-ldapauthenticator</li>
    <li>python310Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python310Packages.jupyterhub-systemdspawner</li>
    <li>python310Packages.jupyterhub-systemdspawner.dist</li>
    <li>python311Packages.batchspawner</li>
    <li>python311Packages.batchspawner.dist</li>
    <li>python311Packages.jupyterhub-ldapauthenticator</li>
    <li>python311Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python311Packages.jupyterhub-systemdspawner</li>
    <li>python311Packages.jupyterhub-systemdspawner.dist</li>
  </ul>
</details>
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>python310Packages.dockerspawner</li>
    <li>python310Packages.dockerspawner.dist</li>
    <li>python310Packages.jupyterhub</li>
    <li>python310Packages.jupyterhub-tmpauthenticator</li>
    <li>python310Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python310Packages.jupyterhub.dist</li>
    <li>python310Packages.oauthenticator</li>
    <li>python310Packages.oauthenticator.dist</li>
    <li>python311Packages.dockerspawner</li>
    <li>python311Packages.dockerspawner.dist</li>
    <li>python311Packages.jupyterhub</li>
    <li>python311Packages.jupyterhub-tmpauthenticator</li>
    <li>python311Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python311Packages.jupyterhub.dist</li>
    <li>python311Packages.oauthenticator</li>
    <li>python311Packages.oauthenticator.dist</li>
  </ul>
</details>

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 235828` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>python310Packages.batchspawner</li>
    <li>python310Packages.batchspawner.dist</li>
    <li>python310Packages.dockerspawner</li>
    <li>python310Packages.dockerspawner.dist</li>
    <li>python310Packages.jupyterhub</li>
    <li>python310Packages.jupyterhub-ldapauthenticator</li>
    <li>python310Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python310Packages.jupyterhub-systemdspawner</li>
    <li>python310Packages.jupyterhub-systemdspawner.dist</li>
    <li>python310Packages.jupyterhub-tmpauthenticator</li>
    <li>python310Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python310Packages.jupyterhub.dist</li>
    <li>python310Packages.oauthenticator</li>
    <li>python310Packages.oauthenticator.dist</li>
    <li>python311Packages.batchspawner</li>
    <li>python311Packages.batchspawner.dist</li>
    <li>python311Packages.dockerspawner</li>
    <li>python311Packages.dockerspawner.dist</li>
    <li>python311Packages.jupyterhub</li>
    <li>python311Packages.jupyterhub-ldapauthenticator</li>
    <li>python311Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python311Packages.jupyterhub-systemdspawner</li>
    <li>python311Packages.jupyterhub-systemdspawner.dist</li>
    <li>python311Packages.jupyterhub-tmpauthenticator</li>
    <li>python311Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python311Packages.jupyterhub.dist</li>
    <li>python311Packages.oauthenticator</li>
    <li>python311Packages.oauthenticator.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/batchspawner/default.nix [hunk 1]
```diff
@@ -32,8 +34,9 @@ buildPythonPackage rec {
 
   meta = with lib; {
     description = "A spawner for Jupyterhub to spawn notebooks using batch resource managers";
-    homepage = "https://jupyter.org";
+    homepage = "https://github.com/jupyterhub/batchspawner";
+    changelog = "https://github.com/jupyterhub/batchspawner/blob/v${version}/CHANGELOG.md";
     license = licenses.bsd3;
-    maintainers = [ maintainers.costrouc ];
+    maintainers = with maintainers; [ costrouc ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/batchspawner/default.nix:40]:
```suggestion
    maintainers = with maintainers; [ ];
```
that maintainer is rather inactive and I'd rather be honest than to lie about the maintenance of this package

File: pkgs/development/python-modules/jupyterhub-systemdspawner/default.nix [hunk 2]
```diff
@@ -41,9 +47,14 @@ buildPythonPackage rec {
...
+  pythonImportsCheck = [
+    "systemdspawner"
+  ];
+
   meta = with lib; {
     description = "JupyterHub Spawner using systemd for resource isolation";
     homepage = "https://github.com/jupyterhub/systemdspawner";
+    changelog = "https://github.com/jupyterhub/systemdspawner/blob/v${version}/CHANGELOG.md";
     license = licenses.bsd3;
     maintainers = with maintainers; [ costrouc ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jupyterhub-systemdspawner/default.nix:59]:
```suggestion
    maintainers = with maintainers; [ ];
```

File: pkgs/development/python-modules/jupyterhub/default.nix [hunk 3]
```diff
@@ -151,14 +168,37 @@ buildPythonPackage rec {
     "test_external_service"
     # attempts to do ssl connection
     "test_connection_notebook_wrong_certs"
+    # AttributeError: 'coroutine' object...
+    "test_valid_events"
+    "test_invalid_events"
+    "test_user_group_roles"
+  ];
+
+  disabledTestPaths = [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jupyterhub/default.nix:177]: This should get a comment

**Comment by DanielKriz** [comment]
I would like to ask about the state of this PR as I am depending on it. Is the contributor still active?

File: pkgs/development/python-modules/jupyterhub/default.nix (hunk 3 above)
Comment (fabaff) [reply pkgs/development/python-modules/jupyterhub/default.nix:177]: Comment added.

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 235828` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>python310Packages.batchspawner</li>
    <li>python310Packages.batchspawner.dist</li>
    <li>python310Packages.dockerspawner</li>
    <li>python310Packages.dockerspawner.dist</li>
    <li>python310Packages.jupyterhub</li>
    <li>python310Packages.jupyterhub-ldapauthenticator</li>
    <li>python310Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python310Packages.jupyterhub-systemdspawner</li>
    <li>python310Packages.jupyterhub-systemdspawner.dist</li>
    <li>python310Packages.jupyterhub-tmpauthenticator</li>
    <li>python310Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python310Packages.jupyterhub.dist</li>
    <li>python310Packages.oauthenticator</li>
    <li>python310Packages.oauthenticator.dist</li>
    <li>python311Packages.batchspawner</li>
    <li>python311Packages.batchspawner.dist</li>
    <li>python311Packages.dockerspawner</li>
    <li>python311Packages.dockerspawner.dist</li>
    <li>python311Packages.jupyterhub</li>
    <li>python311Packages.jupyterhub-ldapauthenticator</li>
    <li>python311Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python311Packages.jupyterhub-systemdspawner</li>
    <li>python311Packages.jupyterhub-systemdspawner.dist</li>
    <li>python311Packages.jupyterhub-tmpauthenticator</li>
    <li>python311Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python311Packages.jupyterhub.dist</li>
    <li>python311Packages.oauthenticator</li>
    <li>python311Packages.oauthenticator.dist</li>
  </ul>
</details>

**Comment by fabaff** [comment]
@ofborg build python311Packages.jupyterhub python311Packages.batchspawner python311Packages.oauthenticator

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 235828` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>python310Packages.batchspawner</li>
    <li>python310Packages.batchspawner.dist</li>
    <li>python310Packages.dockerspawner</li>
    <li>python310Packages.dockerspawner.dist</li>
    <li>python310Packages.jupyterhub</li>
    <li>python310Packages.jupyterhub-ldapauthenticator</li>
    <li>python310Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python310Packages.jupyterhub-systemdspawner</li>
    <li>python310Packages.jupyterhub-systemdspawner.dist</li>
    <li>python310Packages.jupyterhub-tmpauthenticator</li>
    <li>python310Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python310Packages.jupyterhub.dist</li>
    <li>python310Packages.oauthenticator</li>
    <li>python310Packages.oauthenticator.dist</li>
    <li>python311Packages.batchspawner</li>
    <li>python311Packages.batchspawner.dist</li>
    <li>python311Packages.dockerspawner</li>
    <li>python311Packages.dockerspawner.dist</li>
    <li>python311Packages.jupyterhub</li>
    <li>python311Packages.jupyterhub-ldapauthenticator</li>
    <li>python311Packages.jupyterhub-ldapauthenticator.dist</li>
    <li>python311Packages.jupyterhub-systemdspawner</li>
    <li>python311Packages.jupyterhub-systemdspawner.dist</li>
    <li>python311Packages.jupyterhub-tmpauthenticator</li>
    <li>python311Packages.jupyterhub-tmpauthenticator.dist</li>
    <li>python311Packages.jupyterhub.dist</li>
    <li>python311Packages.oauthenticator</li>
    <li>python311Packages.oauthenticator.dist</li>
  </ul>
</details>

**Pull request merged**